c++/src/kj/thread.c++

```diff
@@ -102,18 +107,37 @@ void Thread::detach() {
     KJ_FAIL_SYSCALL("pthread_detach", pthreadResult) { break; }
   }
   detached = true;
+  state->unref();
 }
 
 void* Thread::runThread(void* ptr) {
-  Thread* thread = reinterpret_cast<Thread*>(ptr);
+  ThreadState* state = reinterpret_cast<ThreadState*>(ptr);
   KJ_IF_MAYBE(exception, kj::runCatchingExceptions([&]() {
-    thread->func();
+    state->func();
   })) {
-    thread->exception = kj::mv(*exception);
+    state->exception = kj::mv(*exception);
   }
+  state->unref();
   return nullptr;
 }
 
 #endif  // _WIN32, else
 
+void Thread::ThreadState::unref() {
+#if _MSC_VER
+  if (_InterlockedDecrement_rel(&refcount)) {
+    _ReadBarrier();
+#else
+  if (__atomic_sub_fetch(&refcount, 1, __ATOMIC_RELEASE) == 0) {
```

Comment:
I thought this was gcc specific but I can confirm it works with OSX+clang.  I don't know if the `<stdatomic.h>` versions are an option to widen your supported compilers, but your listed supported compilers are covered.